miner/env_changes.go

```diff
@@ -0,0 +1,417 @@
+package miner
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// envChanges is a helper struct to apply and revert changes to the environment
+type envChanges struct {
+	env      *environment
+	gasPool  *core.GasPool
+	usedGas  uint64
+	profit   *big.Int
+	txs      []*types.Transaction
+	receipts []*types.Receipt
+}
+
+func newEnvChanges(env *environment) (*envChanges, error) {
+	if err := env.state.MultiTxSnapshot(); err != nil {
+		return nil, err
+	}
+
+	return &envChanges{
+		env:      env,
+		gasPool:  new(core.GasPool).AddGas(env.gasPool.Gas()),
+		usedGas:  env.header.GasUsed,
+		profit:   new(big.Int).Set(env.profit),
+		txs:      make([]*types.Transaction, 0),
+		receipts: make([]*types.Receipt, 0),
+	}, nil
+}
+
+func (c *envChanges) commitPayoutTx(
+	amount *big.Int, sender, receiver common.Address,
+	gas uint64, prv *ecdsa.PrivateKey, chData chainData) (*types.Receipt, error) {
+	return commitPayoutTx(PayoutTransactionParams{
+		Amount:        amount,
+		BaseFee:       c.env.header.BaseFee,
+		ChainData:     chData,
+		Gas:           gas,
+		CommitFn:      c.commitTx,
+		Receiver:      receiver,
+		Sender:        sender,
+		SenderBalance: c.env.state.GetBalance(sender),
+		SenderNonce:   c.env.state.GetNonce(sender),
+		Signer:        c.env.signer,
+		PrivateKey:    prv,
+	})
+}
+
+func (c *envChanges) commitTx(tx *types.Transaction, chData chainData) (*types.Receipt, int, error) {
+	var (
+		gasPoolBefore  = new(core.GasPool).AddGas(c.gasPool.Gas())
+		usedGasBefore  = c.usedGas
+		txsBefore      = c.txs[:]
+		receiptsBefore = c.receipts[:]
+		profitBefore   = new(big.Int).Set(c.profit)
+	)
+	signer := c.env.signer
+	sender, err := types.Sender(signer, tx)
+	if err != nil {
+		return nil, popTx, err
+	}
+
+	gasPrice, err := tx.EffectiveGasTip(c.env.header.BaseFee)
+	if err != nil {
+		return nil, shiftTx, err
+	}
+
+	if _, in := chData.blacklist[sender]; in {
+		return nil, popTx, errors.New("blacklist violation, tx.sender")
+	}
+
+	if to := tx.To(); to != nil {
+		if _, in := chData.blacklist[*to]; in {
+			return nil, popTx, errors.New("blacklist violation, tx.to")
+		}
+	}
+
+	cfg := *chData.chain.GetVMConfig()
+	touchTracer := logger.NewAccountTouchTracer()
```

Comment:
unfortunately account touch tracer is broken. we should just remove and use whatever is used for block validation api (access list tracer). thats my bad, I should have just removed that from the codebase

Reply:
no problem, appreciate the context - updated to use access list tracer